### PR TITLE
GEOMESA-1608 Implement/test SparkSQL spatial relationship functions

### PR DESCRIPTION
**Pull request by mdzimmerman**
* Implements st_area, st_distance, st_length, st_closestpoint, st_relate
* More thorough tests of DE-9IM relationships (tests both Spark level
  and passthrough to GeoTools store)
* Adds an option to the geomesa-cqengine-datastore to turn off geo-index

Signed-off-by: Matthew Zimmerman <matt.zimmerman@ccri.com>

### CODE REVIEW

File: geomesa-memory/geomesa-cqengine/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala [hunk 1]
```diff
@@ -32,14 +32,15 @@ import org.opengis.filter._
...
 class GeoCQEngine(val sft: SimpleFeatureType,
-                  enableFidIndex: Boolean = false) extends LazyLogging {
+                  enableFidIndex: Boolean = false,
+                  enableGeomIndex: Boolean = true) extends LazyLogging {
   //val cqcache = CQIndexingOptions.buildIndexedCollection(sft)
   val cqcache: IndexedCollection[SimpleFeature] = new ConcurrentIndexedCollection[SimpleFeature]()
   val attributes = SFTAttributes(sft)
 
   // Add Geometry index on default geometry first.
   // TODO: Add logic to allow for the geo-index to be disabled?  (Low priority)
```
Comment (aheyne) [inline geomesa-memory/geomesa-cqengine/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala]: Is this comment still relevant?

File: geomesa-memory/geomesa-cqengine-datastore/src/main/scala/org/locationtech/geomesa/memory/cqengine/datastore/GeoCQEngineDataStoreFactory.scala [hunk 2]
```diff
@@ -30,7 +34,30 @@ class GeoCQEngineDataStoreFactory extends DataStoreFactorySpi {
...
+    UseGeoIndexParam
+  )
+
+  def getUseGeoIndex(params: util.Map[String, Serializable]): Boolean = {
+    if (params.containsKey(UseGeoIndexKey))
+      UseGeoIndexParam.lookUp(params).asInstanceOf[Boolean]
+    else
+      UseGeoIndexDefault
+  }
+}
```
Comment (aheyne) [inline geomesa-memory/geomesa-cqengine-datastore/src/main/scala/org/locationtech/geomesa/memory/cqengine/datastore/GeoCQEngineDataStoreFactory.scala]: New Line

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLSpatialFunctions.scala [hunk 3]
```diff
@@ -58,9 +69,17 @@ object SQLSpatialFunctions {
...
-
-    sqlContext.udf.register("st_centroid"      , ST_Centroid)
-    sqlContext.udf.register("st_distanceSpheroid"  , ST_DistanceSpheroid)
+    // renamed st_relate variant that returns a boolean since
+    // Spark SQL doesn't seem to support polymorphic UDFs
+    sqlContext.udf.register("st_relate"      , ST_Relate)
+    sqlContext.udf.register("st_relateBool"  , ST_RelateBool)
+
+    sqlContext.udf.register("st_area"            , ST_Area)
+    sqlContext.udf.register("st_closestpoint"    , ST_ClosestPoint)
```
Comment (aheyne) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLSpatialFunctions.scala]: Are you following camel case here?

File: geomesa-spark/geomesa-spark-sql/src/test/scala/org/locationtech/geomesa/spark/SparkSQLSpatialRelationshipsTest.scala [hunk 4]
```diff
@@ -26,171 +28,360 @@ class SparkSQLSpatialRelationshipsTest extends Specification with LazyLogging {
...
-        r.head.getBoolean(0) mustEqual expected
-      }
+    def testData(sql: String, expectedNames: Seq[String]) = {
+      val r = sc.sql(sql)
+      val d = r.collect()
+      val column = d.map(row => row.getAs[String]("name")).toSeq
+      column must containTheSameElementsAs(expectedNames)
+    }
+
+    def testDirect(f: String, name: String, g1: String, g2: String, expected: Boolean) = {
```
Comment (aheyne) [inline geomesa-spark/geomesa-spark-sql/src/test/scala/org/locationtech/geomesa/spark/SparkSQLSpatialRelationshipsTest.scala]: Probably should have an inner test block here that prints the name and function.

File: geomesa-memory/geomesa-cqengine/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala (hunk 1 above)
Comment (anthonyccri) [inline geomesa-memory/geomesa-cqengine/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala]: Done

**Comment by anthonyccri** [comment]
Merging

**Comment by anthonyccri** [comment]
Merged

**Pull request closed without merging**